### PR TITLE
FEAT: create gpu feature on dashboard

### LINKED ISSUE #253: BUG: incorrect memory info when GPU is available
### Describe the bug
When GPU is available, the amount of memory showed in the dashboard is not correct. Looks like the GPU memory is included.

Also, the resource page cannot be accessed.

### To Reproduce
Create an Xorbits cluster on a host with GPU.

1. Your Python version
3.9.15
2. The version of Xorbits you use
0.2.0+5.gf204d77

### PR DESCRIPTION
**Pull request by RayJi01**
create gpu usage and gpu memory usage column on xorbits web dashboard.

Previously, the dashboard of our websites only showing the CPU and usage and CPU memory usage. However, the CPU memory usage is actually calculating the sum of GPU and CPU memory usage. This update distinguishes the CPU and GPU memory usage and adds the GPU information on dashboard. Now the dashboard will display CPU usage, CPU memory usage, GPU usage, and GPU memory usage. 

Resolve #253 

## Check code requirements

- [x] tests added / passed (if needed)
- [x] Ensure all linting tests pass

### CODE REVIEW

File: Miniconda3-latest-Linux-x86_64.sh
Comment (UranusSeven) [inline Miniconda3-latest-Linux-x86_64.sh:1]: Seems that you added some files accidentally to your pull request.

**Comment by ChengjieLi28** [comment]
Can you please attach some screenshots on this pull request to let us see the frontend changes?

**Comment by ChengjieLi28** [comment]
Regarding GPU information, do you think it would be best if we don't display it when users are using the UI on a machine without a GPU card? What do you think?

**Comment by RayJi01** [comment]
image

**Comment by RayJi01** [comment]
f4040161-0e28-4b52-b0d8-a8e8c593e9cd

**Comment by RayJi01** [comment]
After running 3 times of print(np.sort(np.random.rand(100_000_000, gpu=True))):
The usage have been changed:
image

**Comment by RayJi01** [comment]
I believe if users don't have to GPU, then our backend code will return an default value which should represents that their gpu usage is dummy, which should not be too misleading.

File: python/xorbits/web/ui/src/Dashboard.js [hunk 1]
```diff
@@ -86,73 +86,189 @@ class NodeInfo extends React.Component {
...
-            </StyledTableCell>
-          </StyledTableRow>
-          <StyledTableRow>
-            <StyledTableCell>Memory Info</StyledTableCell>
-            <StyledTableCell>
-              <Grid container>
-                <Grid xs={4}>
-                  Usage: {toReadableSize(resourceStats.memory_used)}
+    //in case that we have no gpu in use.
+    if (resourceStats.gpu_memory_total === undefined) {
```
Comment (ChengjieLi28) [inline python/xorbits/web/ui/src/Dashboard.js:111]: Just ``if (resourceStats.gpu_memory_total)`` is fine for javascript.

File: python/xorbits/web/ui/src/Dashboard.js [hunk 2]
```diff
@@ -86,73 +86,189 @@ class NodeInfo extends React.Component {
...
-              </Grid>
-            </StyledTableCell>
-          </StyledTableRow>
-        </TableBody>
-      </Table>
-    )
+              </StyledTableCell>
+            </StyledTableRow>
+            <StyledTableRow>
+              <StyledTableCell>CPU Info</StyledTableCell>
```
[the remark this answers is not in the document]
Comment (RayJi01) [reply python/xorbits/web/ui/src/Dashboard.js]: updated in the latest refactor version

File: python/xorbits/web/ui/src/Dashboard.js [hunk 3]
```diff
@@ -86,73 +86,189 @@ class NodeInfo extends React.Component {
         )
       )
 
+    let target = Object.values(roleData)
+    let res = target.at('resource')
+    console.log(target)
+    console.log(res)
```
[the remark this answers is not in the document]
Comment (RayJi01) [reply python/xorbits/web/ui/src/Dashboard.js]: updated in the latest refactor version

**Comment by RayJi01** [comment]
Most recent update on May 31st:

Update and fix the Workers and Supervisor -> resources page on the dashboard. Here are the result running with GPU available machine:

e15bfd06-7ebc-4250-9c2c-7f845a13d457
da177ba5-5098-4a89-bcda-5838797a2ace

File: python/xorbits/_mars/lib/nvutils.py [hunk 4]
```diff
@@ -535,7 +535,6 @@ def get_handle_by_uuid(uuid: bytes) -> _nvmlDevice_t:
 
 
 def get_mig_mode(device: _nvmlDevice_t) -> Tuple[int, int]:
-    _init_nvml()
```
Comment (qianduoduo0904) [inline python/xorbits/_mars/lib/nvutils.py:538]: Why removing this line?

File: python/xorbits/web/ui/src/node_info/NodeResourceTab.js [hunk 5]
```diff
@@ -61,25 +61,74 @@ export default class NodeResourceTab extends React.Component {
...
       let detail = this.state.resource[band]
-      converted[band] = {
-        CPU: {},
-        Memory: {},
-      }
-      let cpuAvail = detail['cpu_avail']
-      let cpuTotal = detail['cpu_total']
-      let cpuUsage = (cpuTotal - cpuAvail).toFixed(2)
+      console.log(detail)
+      console.log(band)
```
Comment (ChengjieLi28) [inline python/xorbits/web/ui/src/node_info/NodeResourceTab.js:89]: Remove console output.

File: python/xorbits/web/ui/src/node_info/NodeResourceTab.js [hunk 6]
```diff
@@ -61,25 +61,74 @@ export default class NodeResourceTab extends React.Component {
...
+      if (band === 'numa-0') {
+        converted[band] = {
+          CPU: {},
+          Memory: {},
+        }
+
+        let cpuAvail = detail['cpu_avail']
+        let cpuTotal = detail['cpu_total']
+
+        console.log(cpuAvail)
```
Comment (ChengjieLi28) [inline python/xorbits/web/ui/src/node_info/NodeResourceTab.js:99]: Ditto.

File: python/xorbits/web/ui/src/node_info/NodeResourceTab.js [hunk 7]
```diff
@@ -61,25 +61,74 @@ export default class NodeResourceTab extends React.Component {
 
   generateBandRows() {
     const converted = {}
+
+    const tablecontentReturn = (converted, band) => {
```
Comment (ChengjieLi28) [inline python/xorbits/web/ui/src/node_info/NodeResourceTab.js:65]: Should this be ``tableContentReturn`` (Hump Naming) ? And just ``tableContent`` is fine.

File: python/xorbits/web/ui/src/Dashboard.js (hunk 1 above)
Comment (ChengjieLi28) [reply python/xorbits/web/ui/src/Dashboard.js:164]: Why this has to be ``=== undefined`` ?

File: python/xorbits/web/ui/src/node_info/NodeResourceTab.js [hunk 8]
```diff
@@ -61,25 +61,74 @@ export default class NodeResourceTab extends React.Component {
...
+
+        converted[band]['CPU']['CpuUsage'] = cpuUsage
+        converted[band]['CPU']['CpuTotal'] = cpuTotal.toFixed(2)
 
-      let memAvail = detail['memory_avail']
-      let memTotal = detail['memory_total']
-      let memUsage = toReadableSize(memTotal - memAvail)
+        converted[band]['Memory']['MemUsage'] = memUsage
+        converted[band]['Memory']['MemTotal'] = toReadableSize(memTotal)
+      } else if (band === 'gpu-0') {
```
Comment (ChengjieLi28) [inline python/xorbits/web/ui/src/node_info/NodeResourceTab.js:111]: Why always ``gpu-0``? If there are multiple gpus on the machine, what's the situation?

File: python/xorbits/web/ui/src/Dashboard.js (hunk 1 above)
Comment (RayJi01) [reply python/xorbits/web/ui/src/Dashboard.js]: resolved in latest version

File: python/xorbits/_mars/lib/nvutils.py (hunk 4 above)
Comment (RayJi01) [reply python/xorbits/_mars/lib/nvutils.py]: seems like a merging issue, fixed in the latest commit

File: python/xorbits/web/ui/src/node_info/NodeResourceTab.js (hunk 7 above)
Comment (RayJi01) [reply python/xorbits/web/ui/src/node_info/NodeResourceTab.js]: refactored in the latest version

File: python/xorbits/web/ui/src/node_info/NodeResourceTab.js (hunk 8 above)
Comment (RayJi01) [reply python/xorbits/web/ui/src/node_info/NodeResourceTab.js]: try to fix in the latest version, pending verification.

File: python/xorbits/web/ui/src/node_info/NodeResourceTab.js (hunk 5 above)
Comment (RayJi01) [reply python/xorbits/web/ui/src/node_info/NodeResourceTab.js]: cleared

**Comment by RayJi01** [comment]
Latest update on June 1st:

Resolved the multiple GPU cases:

When the user uses machines with more than 1 GPU, our website will always support every display reflecting every GPU resource usage. Here are the results:

When running with Machine in Hangzhou HQ with 2 GPU cards present, after calling xorbits.init()
image
image

And then calling:
   import xorbits.numpy as np
   a = np.random.randint(100, size=1000000, chunk_size=100000, gpu=True)
   print(a)
The GPU displays indicate the changes in usage as shown below:
image
image

The results should verify that now our website will support multiple GPU cases.

File: python/xorbits/web/ui/src/node_info/NodeResourceTab.js (hunk 8 above)
Comment (RayJi01) [reply python/xorbits/web/ui/src/node_info/NodeResourceTab.js]: resolved in the latest edition, see the most recent comment below for details.

**Comment by ChengjieLi28** [comment]
Please add some screenshots for this case: There are two GPUs  on the machine, but user only uses one of them.
``xorbits.init(cuda_devices=[1])``

**Comment by RayJi01** [comment]
> Please add some screenshots for this case: There are two GPUs on the machine, but user only uses one of them. `xorbits.init(cuda_devices=[1])`

When running with only the "second" GPU (1 means the second in GPU) in the init process, the backend still transfers all the GPU information to the front end.

Yet when we run further computations as below:

In [5]: import xorbits.numpy as np

In [6]: a = np.random.randint(100, size=1000000, chunk_size=100000, gpu=True)

In [7]: a

Only the GPU-1 usage shows differences, which is exactly the indicated GPU in our cases. The results are shown below:
8fbde67d-0416-44d6-ad17-bf5e1debeee9

**Comment by ChengjieLi28** [comment]
>

**Comment by ChengjieLi28** [comment]
> > Please add some screenshots for this case: There are two GPUs on the machine, but user only uses one of them. `xorbits.init(cuda_devices=[1])`
> 
> When running with only the "second" GPU (1 means the second in GPU) in the init process, the only GPU shows up now, regardless of there are 2 GPU in the machines, is GPU-1, the one we indicated: image
> 
> When we run further computations as below:
> 
> In [5]: import xorbits.numpy as np In [6]: a = np.random.randint(100, size=1000000, chunk_size=100000, gpu=True) In [7]: a
> 
> the results shown like this: image
> 
> The supervisor page remains shows every GPU available for users, regardless of users using or not. The results are here. NOTICE that when user want to only use GPU-1, the GPU-1 is still the only one to be modified. image

Supervisor never uses GPU. So I think GPU-related info can be only showed on Worker page. What do you think?

**Comment by RayJi01** [comment]
> > > Please add some screenshots for this case: There are two GPUs on the machine, but user only uses one of them. `xorbits.init(cuda_devices=[1])`
> > 
> > 
> > When running with only the "second" GPU (1 means the second in GPU) in the init process, the only GPU shows up now, regardless of there are 2 GPU in the machines, is GPU-1, the one we indicated: image
> > When we run further computations as below:
> > In [5]: import xorbits.numpy as np In [6]: a = np.random.randint(100, size=1000000, chunk_size=100000, gpu=True) In [7]: a
> > the results shown like this: image
> > The supervisor page remains shows every GPU available for users, regardless of users using or not. The results are here. NOTICE that when user want to only use GPU-1, the GPU-1 is still the only one to be modified. image
> 
> Supervisor never uses GPU. So I think GPU-related info can be only showed on Worker page. What do you think?

Updated in the newest version: 

This is the results running on AWS machine with simple GPU presents. 
image
51000629-770d-4fcf-b499-ccb2f127816b

File: python/xorbits/web/ui/src/node_info/NodeResourceTab.js [hunk 9]
```diff
@@ -37,6 +37,7 @@ export default class NodeResourceTab extends React.Component {
 
   refreshInfo() {
     fetch(
+      // eslint-disable-next-line react/prop-types
```
Comment (ChengjieLi28) [inline python/xorbits/web/ui/src/node_info/NodeResourceTab.js:40]: Is this line neccessary?

File: python/xorbits/web/ui/src/node_info/NodeResourceTab.js [hunk 10]
```diff
@@ -61,25 +62,75 @@ export default class NodeResourceTab extends React.Component {
...
+        let memAvail = detail['memory_avail']
+        let memTotal = detail['memory_total']
+        let memUsage = toReadableSize(memTotal - memAvail)
+
+        converted[band]['CPU']['CpuUsage'] = cpuUsage
+        converted[band]['CPU']['CpuTotal'] = cpuTotal.toFixed(2)
+
+        converted[band]['Memory']['MemUsage'] = memUsage
+        converted[band]['Memory']['MemTotal'] = toReadableSize(memTotal)
+        // eslint-disable-next-line react/prop-types
```
Comment (ChengjieLi28) [inline python/xorbits/web/ui/src/node_info/NodeResourceTab.js:109]: Ditto.

**Review by ChengjieLi28 — APPROVED** [review]

**Comment by aresnow1** [comment]
GPU usage shows what data? Based on screenshots, it has been 0 all the time.

**Comment by RayJi01** [comment]
> GPU usage shows what data? Based on screenshots, it has been 0 all the time.

It seems like it will never turn to 1 until GPU is fully used. The data conveys from the backend always shows 0. I believe it may be a minor issue on the backend that deserves a new issue addressing it!

**Comment by aresnow1** [comment]
> > GPU usage shows what data? Based on screenshots, it has been 0 all the time.
> 
> It seems like it will never turn to 1 until GPU is fully used. The data conveys from the backend always shows 0. I believe it may be a minor issue on the backend that deserves a new issue addressing it!

I think for GPU, displaying the amount of memory usage is sufficient. The usage value does not seem to be valuable.

**Comment by qianduoduo0904** [comment]
So open an issue about the GPU usage, we can address it in another PR.

**Comment by RayJi01** [comment]
> So open an issue about the GPU usage, we can address it in another PR.

A New issue has already been submitted with an example screenshot.

**Review by aresnow1 — APPROVED** [review]
LGTM

**Pull request merged**